src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
     {
-        top       = head->bbNext;
-        exit      = nullptr;
-        exitCount = 0;
+        // Make sure we've renumbered such that the bitsets can hold all the bits
+        assert(comp->fgBBNumMax <= comp->fgCurBBEpochSize);
+    }
+
+    //------------------------------------------------------------------------
+    // findLoop: Search for a loop with the given HEAD block.
+    //
+    // Arguments:
+    //    head - Block to be the HEAD of any loop identified
+    //
+    // Return Value:
+    //    true  - Found a valid loop.
+    //    false - Did not find a valid loop.
+    //
+    // Notes:
+    //    May modify flow graph go make loop compact before returning.
+    //    Will set instance fields to track loop's extent and exits.
+    //
+    bool findLoop(BasicBlock* head)
+    {
+        BasicBlock* top = head->bbNext;
 
         //  Blocks that are rarely run have a zero bbWeight and should
         //  never be optimized here
 
         if (top->bbWeight == BB_ZERO_WEIGHT)
         {
-            continue;
+            return false;
         }
 
-        for (pred = top->bbPreds; pred; pred = pred->flNext)
+        for (flowList* pred = top->bbPreds; pred; pred = pred->flNext)
         {
-            /* Is this a loop candidate? - We look for "back edges", i.e. an edge from BOTTOM
-             * to TOP (note that this is an abuse of notation since this is not necessarily a back edge
-             * as the definition says, but merely an indication that we have a loop there).
-             * Thus, we have to be very careful and after entry discovery check that it is indeed
-             * the only place we enter the loop (especially for non-reducible flow graphs).
-             */
+            if (findLoop(head, top, pred->flBlock))
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    //------------------------------------------------------------------------
+    // recordLoop: Notify the Compiler that a loop has been found.
+    //
+    // Return Value:
+    //    true  - Loop successfully recorded.
+    //    false - Compiler has run out of loop descriptors; loop not recorded.
+    //
+    bool recordLoop()
+    {
+        /* At this point we have a compact loop - record it in the loop table
+        * If we found only one exit, record it in the table too
+        * (otherwise an exit = nullptr in the loop table means multiple exits) */
+
+        BasicBlock* onlyExit = (exitCount == 1 ? lastExit : nullptr);
+        if (comp->optRecordLoop(head, first, top, entry, bottom, onlyExit, exitCount))
+        {
+            // Record the BOTTOM block for future reference before returning.
+            assert(bottom->bbNum < oldBlockMaxNum);
+            BlockSetOps::AddElemD(comp, bottomBlocks, bottom->bbNum);
+            return true;
+        }
+
+        // Unable to record this loop because the loop descriptor table overflowed.
+        return false;
+    }
```

Comment:
I went back and took a closer look.  We compile with `COUNT_LOOPS` `#define`d to `0` by default.  Also, the prior behavior was to never bail out of searching for loops, that was something I was adding in this change (per some [earlier feedback](https://github.com/dotnet/coreclr/pull/13314#discussion_r132538713)).  So I've rearranged things to count all loops when `COUNT_LOOPS` is `1` (as before), but to bail out before finding a doomed loop otherwise.  Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/6d7fcf6d55fcad8b765c14ba215a9f04cd52a4e3 (before refactoring) and https://github.com/dotnet/coreclr/pull/13314/commits/2feecd675c62098c876d88cce7a84119da61b086 (after refactoring)